internal/alb/rs/rules.go

```diff
@@ -3,202 +3,339 @@ package rs
 import (
 	"context"
 	"fmt"
+	"strconv"
 
-	"github.com/kubernetes-sigs/aws-alb-ingress-controller/internal/ingress/annotations/action"
-	"k8s.io/apimachinery/pkg/util/intstr"
-
-	"github.com/aws/aws-sdk-go/aws/awsutil"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/request"
 	"github.com/aws/aws-sdk-go/service/elbv2"
-
-	extensions "k8s.io/api/extensions/v1beta1"
-
+	"github.com/aws/aws-sdk-go/service/elbv2/elbv2iface"
+	"github.com/kubernetes-sigs/aws-alb-ingress-controller/internal/alb/tags"
 	"github.com/kubernetes-sigs/aws-alb-ingress-controller/internal/alb/tg"
+	"github.com/kubernetes-sigs/aws-alb-ingress-controller/internal/albctx"
 	"github.com/kubernetes-sigs/aws-alb-ingress-controller/internal/aws/albelbv2"
-	"github.com/kubernetes-sigs/aws-alb-ingress-controller/internal/ingress/annotations/loadbalancer"
+	"github.com/kubernetes-sigs/aws-alb-ingress-controller/internal/ingress/annotations/action"
 	"github.com/kubernetes-sigs/aws-alb-ingress-controller/internal/ingress/controller/store"
+	"github.com/kubernetes-sigs/aws-alb-ingress-controller/internal/k8s"
+	"github.com/kubernetes-sigs/aws-alb-ingress-controller/pkg/util/log"
+	api "k8s.io/api/core/v1"
+	extensions "k8s.io/api/extensions/v1beta1"
+	"k8s.io/apimachinery/pkg/util/intstr"
 )
 
-type NewCurrentRulesOptions struct {
-	ListenerArn  *string
-	TargetGroups tg.TargetGroups
+// Rule contain the elbv2 rule configuration along with the Ingress backend that it is forwarding to
+type Rule struct {
+	Backend extensions.IngressBackend
+	elbv2.Rule
 }
 
-// NewCurrentRules
-func NewCurrentRules(o *NewCurrentRulesOptions) (Rules, error) {
-	var rs Rules
+// Rules contains the rules for a listener.
+type Rules struct {
+	Ingress *extensions.Ingress
 
-	rules, err := albelbv2.ELBV2svc.DescribeRules(&elbv2.DescribeRulesInput{ListenerArn: o.ListenerArn})
-	if err != nil {
-		return nil, err
-	}
+	ListenerArn  string
+	TargetGroups tg.TargetGroups
+	Rules        []*Rule
+}
 
-	for _, r := range rules.Rules {
-		var svcName string
-		var svcPort intstr.IntOrString
+// NewRules returns a new Rules pointer
+func NewRules(ingress *extensions.Ingress) *Rules {
+	return &Rules{
+		Ingress: ingress,
+	}
+}
 
-		if *r.Actions[0].Type == elbv2.ActionTypeEnumForward {
-			i, tg := o.TargetGroups.FindCurrentByARN(*r.Actions[0].TargetGroupArn)
-			if i < 0 {
-				return nil, fmt.Errorf("failed to find a target group associated with a rule. This should not be possible. Rule: %s, ARN: %s", awsutil.Prettify(r.RuleArn), *r.Actions[0].TargetGroupArn)
+func (rs *Rules) TargetGroupArns() (result []string) {
+	for _, rule := range rs.Rules {
```

Comment:
i don't think we need to reply these arns to know which target group we need.
Since when we build these target groups, we already know it. (this can be changed later when we refactoring more)